### ISSUE TITLE
Rule proposal: Prefer `prefer-scoped-selector`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1429]
Many people don't know the `Element.querySelector()` and `Element.querySelectorAll()` behaves differently than most common JavaScript DOM libraries, [detail](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll#user_notes).

Got the idea [here](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1404#issuecomment-875997939)

## Fail

```js
const inner = element.querySelectorAll('.outer .inner');
```

## Pass

```js
const inner = document.querySelectorAll('.outer .inner');
```

```js
const inner = element.querySelectorAll(':scope .outer .inner');
```

---

I'm not sure about these, they are never going to get wrong result.

```js
const inner = element.querySelectorAll('.a');
```

```js
const inner = element.querySelectorAll('.a + .b');
```

Should we enforce them consistently, or only check [descendant combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Descendant_combinator) and [child combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator)?

**Comment by sindresorhus** [comment]
> By default, querySelectorAll() only verifies that the last element in the selector is within the search scope.

Wut. Who thought that would be a good idea...

**Comment by fisker** [comment]
Haha, so you are the one don't know that. It's more like `document.querySelectorAll().filter(inElement)`

**Comment by fregante** [comment]
> > By default, querySelectorAll() only verifies that the last element in the selector is within the search scope.
> 
> Wut. Who thought that would be a good idea...

While it queries the children, I'd think the CSS engine takes over with just "does this element match this selector"? That'd explain the `:scope`-less default.

Not a fan of `:scope` but sure, it makes sense.

> only check [descendant combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Descendant_combinator) and [child combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator)?

But I think this rule, especially if autofixable, would require a compliant CSS selector parser (not that there's lack of it, but they're not tiny)

**Comment by sindresorhus** [comment]
This is accepted.

**Comment by fregante** [comment]
> require a compliant CSS selector parser

Probably not, as it just has to trigger if it lacks `\b:scope\b`, which isn't exactly common. Even if it finds it, it would be a false positive (i.e. no report) so it's less annoying.
